### ISSUE TITLE
Deserialization of DTO input for custom operation 

### ISSUE THREAD

**Issue by yivi** [api-platform/api-platform#1349]
**Description**  
I'm not entirely sure if this is a bug report (which I imagine unlikely), a feature request, or simply a misunderstanding of the documentation on my part.

**Example**  
I'm trying a little custom operation over an existing api resource, which should receive a DTO as input parameter.

I've described the operation like this:

```yml
App\Domain\Entity\Invoice:
  collectionOperations:
    credit_note:
      input: App\Application\Invoice\CreateCreditNoteRequest
      method: POST
      path: /invoices/credit_note
      controller: App\Infrastructure\Http\ApiResource\CreateCreditNote
```

This seems to be picked up correctly by the doc generator, and it shows me the structure for `CreateCreditNoteRequest` in the new action:

<img width="354" alt="image" src="https://user-images.githubusercontent.com/1815039/69903819-ff777200-139e-11ea-93a6-86b70a85619a.png">

However, when submitting the request, it fails because it attempts to deserialize the payload into 
`App\Domain\Entity\Invoice` instead of `App\Application\Invoice\CreateCreditNoteRequest`.

I've trying debugging the process, and I've gotten to `DeserializeListener:107`, that says:

```php
$request->attributes->set(
    'data',
    $this->serializer->deserialize($request->getContent(), $context['resource_class'], $format, $context)
);
```

So no matter what input we declared, it will always try to deserialize to the resource class.

As an experiment, I've changed it to:

```php
 $request->attributes->set(
     'data',
     $this->serializer->deserialize(
         $request->getContent(),
         $context['input']['class'] ?? $context['resource_class'],
         $format,
         $context)
  );
```

(Attempt to deserialize to `input.class` if declared, or continue with `resource_class` otherwise).

Is there another way to accomplish this? If not, wouldn't the above be more practical to handle this scenario? The alternative I've found is to disable deserialization altogether, and handle it myself, but it doesn't smell very good.

**Comment by yivi** [comment]
The question is framed as about "custom operations", but it is the same for regular operations.

If I set the input class for the `post` operation of resource `Foo` to `FooDto`, when I get to `DeserializeListener` the value for `$context['resource_class']` is `Foo`, and it will attempt to deserialize into `Foo` and not into `FooDto`.

Which reinforces the idea that's it's me that I'm doing something wrong. I feel uncertain of submitting a PR for such a small change, I'm sure this is just me misconfiguring things.

**Comment by soyuka** [comment]
> However, when submitting the request, it fails because it attempts to deserialize the payload into
App\Domain\Entity\Invoice instead of App\Application\Invoice\CreateCreditNoteRequest.

Did you implement a data transformer? It should indeed use the `CreateCreditNoteRequest` on this `credit_note` POST operation.

**Comment by soyuka** [comment]
This test file may help you: https://github.com/api-platform/core/blob/26cfe160ccfc1d083878bc65710efb45d6dac924/features/jsonld/input_output.feature

**Comment by yivi** [comment]
Hmmm... do I **have** to create Data transformer? Or are you asking that because the existence of a Data transformer could have messed the DTO deserialization?

In any case, we have hone on this project.

But if you say it _should_ work as is, and it's not working, I guessed we messed up somewhere else.

What I found weird is that by making the change I mentioned above I did get it to work, but obviously I'm missing part of the picture.

I'll investigate a bit more and try to either create a reproducible sample, or to find the issue and close this.

Thanks.

**Comment by yivi** [comment]
I created a project from scratch (`create-project symfony/skeleton here "4.*" && composer require api`), and tried to replicate the problem.

And I get the same results.

This project is blank except two "entity" classes, one dummy controller, and one configuration file to define the resource and the operations.

It's "all" [here](https://github.com/yivi/api-platform-dto-mcve).

To keep it neat, it's runnable by doing `make build cinstall cautoload serve`, and you'd get it on `http://localhost:8000/api`.

I get the same results than in the real project:
- Api docs are correctly generated (uses the Dto to document both the request body and the example response)
- When I execute it, it fails because the controller expects `__invoke(FooDto $data)` but the `class_resource` has been deserialized into `$data` (`VeryRealEntity`, for this project).

Am I doing something wrong? Should I be creating a data transformer?

**Comment by soyuka** [comment]
Hi,

I just tried your example. Indeed, I don't think that `input` / `output` is what you want here, or you're misusing it. 
Could you tell me what the controller is supposed to do? Note that in the case of a `POST` with an `Input` on `Entity` what should happen is:

POST `{input}` => deserialized as `Input` => goes through the `Transformer` (that you need to implement to transform `Input` into `Entity`) => `Entity` => Controller with `__invoke(Entity $entity)`.

**Comment by yivi** [comment]
Yeah, I imagined the mistake was on my side.

At high level, this controller creates a "credit note" for an invoice. There is no "CreditNote" resource, since a credit note is simply a specialized kind of invoice.

It receives a DTO with the original invoice Id, to be able to link the newly created credit note and the original invoice, and a few extra fields that are **not** part of the Invoice entity that act as extra configuration for the CreditNoteHandler (tied to payments, notifications, etc).

Re-reading [the docs](https://api-platform.com/docs/core/dto/#specifying-an-input-or-an-output-data-representation), my confusion becomes obvious. As in, I'm not sure how I managed to understand otherwise.

It's clear to me now that to use a DTO I should have implemented a data transformer. That's not exactly what I need for this particular case, since the DTO is a command object that can't be simply transformed into an entity. We'll just keep reading and deserialization for this endpoint, and manage that ourselves using  our own listener for this.

Thanks a lot for your help.
